src/Route.php

```diff
@@ -169,17 +183,58 @@ public function defaults(array $defaults): self
     }
 
     /**
-     * @param callable|MiddlewareInterface $middleware
-     * @return MiddlewareInterface
+     * @param MiddlewareInterface|callable|string|array $middleware
      */
-    private function prepareMiddleware($middleware): MiddlewareInterface
+    private function validateMiddleware($middleware): void
     {
+        if (
+            is_string($middleware) && is_subclass_of($middleware, MiddlewareInterface::class)
+        ) {
+            if ($this->container === null) {
+                throw new InvalidArgumentException('Route container must not be null for lazy loaded middleware.');
+            }
+            return;
+        }
+
+        if (
+            is_array($middleware) && count($middleware) === 2
+            && isset($middleware[0], $middleware[1])
+            && is_string($middleware[1]) && is_string($middleware[0]) && class_exists($middleware[0])
+        ) {
+            if ($this->container === null) {
+                throw new InvalidArgumentException('Route container must not be null for handler action.');
+            }
+            return;
+        }
+
         if (is_callable($middleware)) {
-            $middleware = new Callback($middleware);
```

Comment:
The previous check is not necessary now, because is_callable() checks that this is the array with two elements where first is a class name or an object and second it is the class method. But we have to fix the expected error message in the test. What do you think?

Reply:
I think it's good that the messages are specific.